### ISSUE TITLE
Wrong calculation in vouchers created after another voucher with free shipping is partially used

### ISSUE THREAD

**Issue by amaurigabriel** [PrestaShop/PrestaShop#28901]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

Hi,

We opened a support ticket in support.prestashop.com and we were told that we are facing a bug in PrestaShop.
When you have a voucher with free shipping **and** discount value, e.g., 100.00, if the customer create a order with a value smaler than 100.00, a new voucher is created.

This is expected behavior, but the issue is that the new voucher is created with wrong value.

A simple example is:
1. Voucher with free shipping + fixed value of 100.00 in discount
2. Customer create a order with 50.00 in products + 10.00 in shipping
3. The new voucher will be created with a discount of 100.00 - 50.00 - 10.00 - 10.00 (the shipping cost is subtracted twice). The expected value would be 100.00 - 50.00 - 10.00

This seems to be happening because the new voucher value is determined as "origina discount - order total discounts - shipping cost", but the shipping cost is already in "order total discounts".

We reproduced this in PS 1.7.7.2 and 1.7.7.4. PrestaShop support team said they actually reproduced this in 1.7.8.6.

### Expected behavior

In scenario described, the expected discount for new voucher is  100.00 - 50.00 - 10.00 = 40.00.
Current value is 100.00 - 50.00 - 10.00 - 10.00 = 30.00.

### Steps to reproduce

1. Create a voucher with fixed discount value + free shipping
2. Create a order with total value smaller than voucher discount
3. PrestaShop will create a new voucher with incorrect discount

### PrestaShop version(s) where the bug happened

at least 1.7.7.2 to 1.7.8.6

**Comment by hibatallahAouadni** [comment]
Hello @amaurigabriel 

Your issue seems similar to this #28176
And as you can see the QA-team couldn't reproduce the issue :disappointed: 
So, could you please provide more infos :pray: are you using a non-native shipping / payment modules?
Waiting for your feedback.

Thanks!

**Comment by amaurigabriel** [comment]
Hi, @hibatallahAouadni , thank you for your response.

As our issue happens when the order value is **smaller** than the voucher discount, no payment gateway is required (the order is always free).
I recorded a video with the steps to reproduce this in a clean intallation (portuguese language and no taxes in shop are the only two modifications). This is a test shop, so I can share the BackOffice credentials with you if necessary.

[screen-capture.webm](https://user-images.githubusercontent.com/1836086/176558662-4941c79a-958c-40fd-930f-2944980bb093.webm)

Please see that I created a voucher with discount of 1000 + free shipping, and I purchased a order with 29.00 products + 7.00 of shipping = 36.00. So the new voucher should have been created with discount of 1000 - 36 = 964. But, actually, it was created with discount of 100 - 36 - 7 = 957.

** PS. I manually created the carrier using the BackOffice screen.

**Comment by sLorenzini** [comment]
hello @amaurigabriel,

I coudn't reproduce your issue and your video is not working.
Could you provide us another screenrecord please?

Waiting for your feedback.

Thank you

**Comment by amaurigabriel** [comment]
Hi, @sLorenzini 

I opened this issue in anonymous mode and the video worked perfectly. It is just a .webm video, any major browser should be able to reproduce it.

https://caniuse.com/webm

![image](https://user-images.githubusercontent.com/1836086/176654519-d5e1e7f2-1de0-42e1-b55f-dd4d1defff79.png)

**Comment by sLorenzini** [comment]
hello @amaurigabriel,

Thank you for your answer.
I reproduce your issue on 1.7.7.2 & 1.7.8.6:

https://user-images.githubusercontent.com/2854834/176683257-e061b65c-99f3-4b4e-a18a-663eda9df811.mp4


But I need an advice from the PrestaShop product team to know if it's the expected behavior or if it's a real issue to fix. @PrestaShop/product-team  what do you think?

Thanks

**Comment by amaurigabriel** [comment]
Hi,

Any position about this?

**Comment by manfield** [comment]
Hello,
I can reproduce this issue even on the latest Prestashop version.
This bug has been there since a while and I can confirm it should be fixed.
It could not been an expected behaviour. When it creates the second discount voucher, from a partial use voucher, it sums twice the shipping cost.

Example:

25€ products + 8€ shipping = 33€ cart
Discount = 100€
EXPECTED RESULT
0€ cart
67€ remaining from the voucher

ACTUAL RESULT
59€ remaining from the voucher
As you can see the shipping cost is added again: 67€ - 8€ (that was already considered once) = 59€

I will open a merge request on this. I will try to solve the issue according to PS guidelines.

Cheers
M

**Comment by manfield** [comment]
Yes, I think the problem is in this old piece of code in _classes/PaymentModule.php_ 
on line 1182

```
                if ($voucher->reduction_tax) {
                    // Add total shipping amount only if reduction amount > total shipping
                    if ($voucher->free_shipping == 1 && $voucher->reduction_amount >= $order->total_shipping_tax_incl) {
                        $voucher->reduction_amount -= $order->total_shipping_tax_incl;
                    }
                } else {
                    // Add total shipping amount only if reduction amount > total shipping
                    if ($voucher->free_shipping == 1 && $voucher->reduction_amount >= $order->total_shipping_tax_excl) {
                        $voucher->reduction_amount -= $order->total_shipping_tax_excl;
                    }
                }

```

Since  probably the CartRule workflow was changed during 1.7 development this piece of code is unnecessary because the CartRule reduction amount already is calculated considering free shipping (if enabled for that cart rule).
This piece of code subtracts once more the shipping amount. As result the residual CartRule is created with a wrong amount.

**Comment by mflasquin** [comment]
Friendly reminder @PrestaShop/product-team :)

**Comment by PrestaEdit** [comment]
Reproduced inside 8.0.4 and 8.1.0 too, FIY.

**Comment by RosaBenouamer** [comment]
I totally agree @manfield @jolelievre it could be interesting to merged the linked PR if without impact even if the pricing is refactoring. Couldn't it ?

**Comment by florine2623** [comment]
This has been reproduced on `8.1` and `develop` branch.
The shipping is taken into account in the total use of discount.

<img width="1173" alt="Screenshot 2023-07-25 at 16 21 44" src="https://github.com/PrestaShop/PrestaShop/assets/16019289/ac8e3cb0-893a-4137-bb03-140bb409ae0b">

In my case, it should be 60€ instead of 50€.

**Comment by APabisz** [comment]
We are also affected by this issue in a real merchant use case involving gift cards / amount vouchers.

PrestaShop Support confirmed in our ticket that this behavior is a known PrestaShop core bug and recommended adding our use case here to help prioritize the issue.

<img width="804" height="942" alt="Image" src="https://github.com/user-attachments/assets/4db1412a-af57-411f-8eaa-c36792567716" />

## Business impact

This is especially problematic when vouchers are used as gift cards, because customers expect the voucher to behave like a stable monetary balance.

When a voucher has:

- a fixed amount value, for example `1000 EUR`,
- partial use enabled,
- free shipping enabled,

the remaining voucher value is not calculated as a true unused balance. Instead, the first shipping cost is effectively deducted twice, while the newly generated voucher still keeps the free-shipping benefit.

As a result, the real value of the remaining voucher depends on the shipping cost of the next order. This means the customer may either lose or gain value depending on the next shipping method/order.

## Example

Initial voucher:

```text
Voucher amount: 1000 EUR
Partial use: enabled
Free shipping: enabled
```

First order:

```text
Products total: 100 EUR
Shipping cost: 20 EUR
```

Expected remaining balance:

```text
1000 EUR - 100 EUR - 20 EUR = 880 EUR
```

Actual result:

```text
New voucher amount: 860 EUR
Free shipping: still enabled
```

So the calculation behaves like:

```text
1000 EUR - 100 EUR - 20 EUR - 20 EUR = 860 EUR
```

The shipping cost is deducted twice from the monetary part of the voucher.

## Why this is unstable

If the second order has shipping cost `10 EUR`, the customer receives:

```text
860 EUR + 10 EUR free shipping = 870 EUR
```

So the customer loses `10 EUR` compared with the expected `880 EUR` remaining balance.

If the second order has shipping cost `30 EUR`, the customer receives:

```text
860 EUR + 30 EUR free shipping = 890 EUR
```

So the customer gains `10 EUR` compared with the expected `880 EUR` remaining balance.

This makes the remaining gift card balance inconsistent and dependent on future shipping costs.

## Expected behavior

For amount vouchers used as gift cards, the remaining voucher should represent the actual unused monetary balance after the first order:

```text
Original voucher amount
- products paid by voucher
- shipping paid by voucher
= remaining voucher amount
```

In the example above, the new voucher should represent `880 EUR` of remaining value.

It should not lose or gain value depending on the shipping cost of the next order.

## Request

Could this issue be prioritized or the existing PR reviewed again?

There is already a PR linked to this issue: `#31071`. However, I am not sure whether this PR fully covers the whole problem. The same issue may still appear on the final order that exhausts the voucher amount to zero. In that scenario, the voucher may grant free shipping for the entire order without reducing the voucher amount by the shipping cost, because no new partial-use voucher is created afterwards. This means that the fix should also verify the case where the voucher is fully consumed in the current order, not only the case where a new partial-use voucher is generated.

From a merchant perspective, this is not only a minor calculation issue. It directly affects customer balances and can lead to incorrect gift card values.

**Comment by boo-code** [comment]
Measured on 9.2.x. It reproduces, @manfield's diagnosis is right, and #31071 removes exactly the wrong lines. Two extra facts below: why some reproduction attempts failed, and what state that PR is in.

**The shipping is already inside the discount.** `CartRuleCalculator::applyCartRule()` does

```php
if ($cartRule->free_shipping && $withFreeShipping) {
    $initialShippingFees = $this->calculator->getFees()->getInitialShippingFees();
    $this->calculator->getFees()->subDiscountValueShipping($initialShippingFees);
    $cartRuleData->addDiscountApplied($initialShippingFees);
}
```

so by the time `PaymentModule::createOrderCartRules()` computes `$remainingValue = $cartRuleReductionAmountConverted - $values[...]`, the shipping has already been netted out. The block at `classes/PaymentModule.php:1296-1306` then subtracts `$order->total_shipping_tax_incl` a second time.

Driving the real calculator on cart 2 of the demo shop (products 90.90 tax incl, shipping 7.00) with a voucher of 1000, free shipping, partial use, `reduction_tax = 1`:

```
discount applied by the rule = 97.90        (90.90 products + 7.00 shipping)
remainingValue               = 1000.00 - 97.90 = 902.10
guard 902.10 >= 7.00         = TRUE, so the block subtracts
new voucher value            = 902.10 - 7.00 = 895.10       expected 902.10
```

**The last hop matters and it holds.** The subtraction would be harmless if `$order->total_shipping_tax_incl` were zero once the voucher waived the shipping. It is not: `$order->total_shipping_tax_incl` comes from `abs($cart->getOrderTotal(true, Cart::ONLY_SHIPPING, ...))` (`PaymentModule.php:1136`), and measured with the rule applied that still returns **7.00**, not 0.

**Why two QA attempts did not reproduce it.** The block is guarded by `reduction_amount >= total_shipping`, so it only fires when the remainder is at least the shipping cost. On the same cart with a voucher of **100** instead of 1000 the remainder is `100.00 - 97.90 = 2.10`, the guard is false, no subtraction happens and the replacement voucher is correct. A reproduction therefore needs a voucher whose remainder after the order still exceeds the shipping cost — which is exactly the gift-card shape @APabisz describes.

Worth noting too that the comment above those lines reads *"Add total shipping amount only if reduction amount > total shipping"* while the code subtracts. The comment and the code have disagreed for a long time.

**On #31071.** It is open, approved by @tleon on 2023-10-19, one file, and it deletes precisely that block. The block is still present verbatim on `9.2.x`, so the change still applies. What is stale is the branch: `manfield:develop` is **14,086 commits behind** `develop` and has not been touched since 2023, and GitHub reports `mergeable: null` for it. So this is not blocked on the fix being wrong, it is blocked on a rebase — and, given `9.1.x` is gone, on a decision about which branch it should target now.

I have not opened anything competing; the approved PR owns this.

**Comment by manfield** [comment]
I'm realising with the latest comment this PR was not merged yet. Impressive. 2023.
I would say this bug is affecting many PS users in a very bad way
and deserves a rapid resolution. 

@florine2623
